{
 "metadata": {
  "name": "",
  "signature": "sha256:208ca8abed75d920db63107cef88fd1f73e77ad56575f4b01364df786bbca407"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "samples = (('naive','atac',''),\n",
      "           ('naive','atac','foxo1_ko_'),\n",
      "           ('naive','foxo1',''),\n",
      "           ('lcmv_d12','foxo1',''),)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "for i, (cond, seq, breed) in enumerate(samples):\n",
      "    others = samples[:i] + samples[i+1:]\n",
      "    sql = '''select distinct on (p1.id) \n",
      "p1.*, chr.\"name\" as chr_name,\n",
      "p1.tag_count as {}{}_{}_tag_count,\n",
      "'''.format(breed, cond, seq)\n",
      "    selects, joins = [], []\n",
      "    for j, (oth_cond, oth_seq, oth_breed) in enumerate(others):\n",
      "        counter = j + 2\n",
      "        selects.append('''p{counter}.tag_count as {}{}_{}_tag_count,\n",
      "p{counter}.*'''.format(oth_breed, oth_cond, oth_seq, counter=counter))\n",
      "        joins.append('''left outer join \n",
      "chipseq.peak_{}cd8tcell_{}_{} p{counter}\n",
      "on p1.chromosome_id = p{counter}.chromosome_id\n",
      "and p1.start_end && p{counter}.start_end\n",
      "        '''.format(oth_breed, oth_seq, oth_cond, counter=counter))\n",
      "    \n",
      "    sql += ',\\n'.join(selects)\n",
      "    sql += ''',\n",
      "tcf1.tag_count as tcf1_tag_count, tcf1.*\n",
      "from chipseq.peak_{}cd8tcell_{}_{} p1\n",
      "join genome_reference_mm10.chromosome chr \n",
      "on p1.chromosome_id = chr.id\n",
      "'''.format(breed, seq, cond)\n",
      "    sql += ''.join(joins)\n",
      "    sql += '''left outer join \n",
      "chipseq.peak_cd8tcell_tcf1 tcf1\n",
      "on p1.chromosome_id = tcf1.chromosome_id\n",
      "and p1.start_end && tcf1.start_end\n",
      "left outer join genome_reference_mm10.sequence_transcription_region reg\n",
      "on p1.chromosome_id = reg.chromosome_id\n",
      "and p1.start_end && reg.start_site_1000\n",
      "where reg.id is NOT NULL;\n",
      "\n",
      "'''\n",
      "    \n",
      "    print(sql)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "select distinct on (p1.id) \n",
        "p1.*, chr.\"name\" as chr_name,\n",
        "p1.tag_count as naive_atac_tag_count,\n",
        "p2.tag_count as foxo1_ko_naive_atac_tag_count,\n",
        "p2.*,\n",
        "p3.tag_count as naive_foxo1_tag_count,\n",
        "p3.*,\n",
        "p4.tag_count as lcmv_d12_foxo1_tag_count,\n",
        "p4.*,\n",
        "tcf1.tag_count as tcf1_tag_count, tcf1.*\n",
        "from chipseq.peak_cd8tcell_atac_naive p1\n",
        "join genome_reference_mm10.chromosome chr \n",
        "on p1.chromosome_id = chr.id\n",
        "left outer join \n",
        "chipseq.peak_foxo1_ko_cd8tcell_atac_naive p2\n",
        "on p1.chromosome_id = p2.chromosome_id\n",
        "and p1.start_end && p2.start_end\n",
        "        left outer join \n",
        "chipseq.peak_cd8tcell_foxo1_naive p3\n",
        "on p1.chromosome_id = p3.chromosome_id\n",
        "and p1.start_end && p3.start_end\n",
        "        left outer join \n",
        "chipseq.peak_cd8tcell_foxo1_lcmv_d12 p4\n",
        "on p1.chromosome_id = p4.chromosome_id\n",
        "and p1.start_end && p4.start_end\n",
        "        left outer join \n",
        "chipseq.peak_cd8tcell_tcf1 tcf1\n",
        "on p1.chromosome_id = tcf1.chromosome_id\n",
        "and p1.start_end && tcf1.start_end\n",
        "left outer join genome_reference_mm10.sequence_transcription_region reg\n",
        "on p1.chromosome_id = reg.chromosome_id\n",
        "and p1.start_end && reg.start_site_1000\n",
        "where reg.id is NOT NULL;\n",
        "\n",
        "\n",
        "select distinct on (p1.id) \n",
        "p1.*, chr.\"name\" as chr_name,\n",
        "p1.tag_count as foxo1_ko_naive_atac_tag_count,\n",
        "p2.tag_count as naive_atac_tag_count,\n",
        "p2.*,\n",
        "p3.tag_count as naive_foxo1_tag_count,\n",
        "p3.*,\n",
        "p4.tag_count as lcmv_d12_foxo1_tag_count,\n",
        "p4.*,\n",
        "tcf1.tag_count as tcf1_tag_count, tcf1.*\n",
        "from chipseq.peak_foxo1_ko_cd8tcell_atac_naive p1\n",
        "join genome_reference_mm10.chromosome chr \n",
        "on p1.chromosome_id = chr.id\n",
        "left outer join \n",
        "chipseq.peak_cd8tcell_atac_naive p2\n",
        "on p1.chromosome_id = p2.chromosome_id\n",
        "and p1.start_end && p2.start_end\n",
        "        left outer join \n",
        "chipseq.peak_cd8tcell_foxo1_naive p3\n",
        "on p1.chromosome_id = p3.chromosome_id\n",
        "and p1.start_end && p3.start_end\n",
        "        left outer join \n",
        "chipseq.peak_cd8tcell_foxo1_lcmv_d12 p4\n",
        "on p1.chromosome_id = p4.chromosome_id\n",
        "and p1.start_end && p4.start_end\n",
        "        left outer join \n",
        "chipseq.peak_cd8tcell_tcf1 tcf1\n",
        "on p1.chromosome_id = tcf1.chromosome_id\n",
        "and p1.start_end && tcf1.start_end\n",
        "left outer join genome_reference_mm10.sequence_transcription_region reg\n",
        "on p1.chromosome_id = reg.chromosome_id\n",
        "and p1.start_end && reg.start_site_1000\n",
        "where reg.id is NOT NULL;\n",
        "\n",
        "\n",
        "select distinct on (p1.id) \n",
        "p1.*, chr.\"name\" as chr_name,\n",
        "p1.tag_count as naive_foxo1_tag_count,\n",
        "p2.tag_count as naive_atac_tag_count,\n",
        "p2.*,\n",
        "p3.tag_count as foxo1_ko_naive_atac_tag_count,\n",
        "p3.*,\n",
        "p4.tag_count as lcmv_d12_foxo1_tag_count,\n",
        "p4.*,\n",
        "tcf1.tag_count as tcf1_tag_count, tcf1.*\n",
        "from chipseq.peak_cd8tcell_foxo1_naive p1\n",
        "join genome_reference_mm10.chromosome chr \n",
        "on p1.chromosome_id = chr.id\n",
        "left outer join \n",
        "chipseq.peak_cd8tcell_atac_naive p2\n",
        "on p1.chromosome_id = p2.chromosome_id\n",
        "and p1.start_end && p2.start_end\n",
        "        left outer join \n",
        "chipseq.peak_foxo1_ko_cd8tcell_atac_naive p3\n",
        "on p1.chromosome_id = p3.chromosome_id\n",
        "and p1.start_end && p3.start_end\n",
        "        left outer join \n",
        "chipseq.peak_cd8tcell_foxo1_lcmv_d12 p4\n",
        "on p1.chromosome_id = p4.chromosome_id\n",
        "and p1.start_end && p4.start_end\n",
        "        left outer join \n",
        "chipseq.peak_cd8tcell_tcf1 tcf1\n",
        "on p1.chromosome_id = tcf1.chromosome_id\n",
        "and p1.start_end && tcf1.start_end\n",
        "left outer join genome_reference_mm10.sequence_transcription_region reg\n",
        "on p1.chromosome_id = reg.chromosome_id\n",
        "and p1.start_end && reg.start_site_1000\n",
        "where reg.id is NOT NULL;\n",
        "\n",
        "\n",
        "select distinct on (p1.id) \n",
        "p1.*, chr.\"name\" as chr_name,\n",
        "p1.tag_count as lcmv_d12_foxo1_tag_count,\n",
        "p2.tag_count as naive_atac_tag_count,\n",
        "p2.*,\n",
        "p3.tag_count as foxo1_ko_naive_atac_tag_count,\n",
        "p3.*,\n",
        "p4.tag_count as naive_foxo1_tag_count,\n",
        "p4.*,\n",
        "tcf1.tag_count as tcf1_tag_count, tcf1.*\n",
        "from chipseq.peak_cd8tcell_foxo1_lcmv_d12 p1\n",
        "join genome_reference_mm10.chromosome chr \n",
        "on p1.chromosome_id = chr.id\n",
        "left outer join \n",
        "chipseq.peak_cd8tcell_atac_naive p2\n",
        "on p1.chromosome_id = p2.chromosome_id\n",
        "and p1.start_end && p2.start_end\n",
        "        left outer join \n",
        "chipseq.peak_foxo1_ko_cd8tcell_atac_naive p3\n",
        "on p1.chromosome_id = p3.chromosome_id\n",
        "and p1.start_end && p3.start_end\n",
        "        left outer join \n",
        "chipseq.peak_cd8tcell_foxo1_naive p4\n",
        "on p1.chromosome_id = p4.chromosome_id\n",
        "and p1.start_end && p4.start_end\n",
        "        left outer join \n",
        "chipseq.peak_cd8tcell_tcf1 tcf1\n",
        "on p1.chromosome_id = tcf1.chromosome_id\n",
        "and p1.start_end && tcf1.start_end\n",
        "left outer join genome_reference_mm10.sequence_transcription_region reg\n",
        "on p1.chromosome_id = reg.chromosome_id\n",
        "and p1.start_end && reg.start_site_1000\n",
        "where reg.id is NOT NULL;\n",
        "\n",
        "\n"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os\n",
      "dir = '/Users/karmel/Desktop/Projects/GlassLab/Notes_and_Reports/Miscellaneous_Collaborations/Rodrigo_CD8s_2014_09/Promoters/'\n",
      "for cond, seq, breed in samples:\n",
      "    os.mkdir(dir + '{}{}_{}'.format(breed, cond, seq))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}